control-statistics/src/main/java/fi/nls/oskari/control/statistics/user/AddIndicatorDataHandler.java

```diff
@@ -0,0 +1,109 @@
+        }
+        String id = params.getRequiredParam(ActionConstants.PARAM_ID);
+        StatisticalIndicator indicator = datasource.getIndicator(params.getUser(), id);
+        if(indicator == null) {
+            // indicator removed or is not owned by this user
+            throw new ActionParamsException("Requested invalid indicator:" + id);
+        }
+
+        // This is just an addition to the existing model, NOT the whole model
+        JSONObject selectors = params.getHttpParamAsJSON(StatisticsHelper.PARAM_SELECTORS);
+        StatisticalIndicatorDataModel model = StatisticsHelper.getIndicatorDataModel(selectors);
+
+        // add the layer if it doesn't exist yet (only if linked to the datasource)
+        int regionsetId = params.getRequiredParamInt(StatisticsHelper.PARAM_REGIONSET);
+        DatasourceLayer layer = datasource.getSource().getLayers().stream()
+                .filter(x -> x.getMaplayerId() == regionsetId)
+                .findFirst()
+                .orElseThrow(() -> new ActionParamsException("Invalid regionset: " + regionsetId));
+        if(indicator.getLayer(layer.getMaplayerId()) == null) {
+            indicator.addLayer(layer);
+        }
+        try {
+            indicator.setDataModel(model);
+            Map<String, IndicatorValue> data = parseIndicatorData(params.getHttpParam(PARAM_DATA));
+            if (data != null) {
+                datasource.saveIndicatorData(indicator, regionsetId, data, params.getUser());
+            }
+        } catch (Exception ex) {
+            throw new ActionException("Couldn't save data", ex);
+        }
+
+        if (datasource.canCache()) {
+            // TODO: flush/update caches (also metadata and listing)
+            // Not an issue for now since user indicators are not cached and they are are the
+            //  only ones that can be added/edited/removed
+            StatisticsHelper.flushDataFromCache(datasourceId, id, regionsetId, selectors);
+        }
+        JSONObject jobj = new JSONObject();
+        JSONHelper.putValue(jobj, "id", id);
+        ResponseHelper.writeResponse(params,jobj);
+    }
+
+    protected Map<String, IndicatorValue> parseIndicatorData(String data) throws ActionException {
+        if (data == null) {
+            return null;
+        }
+        try {
+            // read json as map with strings and doubles
+            Map<String, Double> values = MAPPER.readValue(data, new TypeReference<Map<String, Double>>(){});
+            // map doubles to IndicatorValueFloat
+            return values.entrySet().stream()
+                    .collect(Collectors
+                            .toMap( e -> e.getKey(),
+                                    e -> new IndicatorValueFloat(e.getValue())));
+        } catch (IOException e) {
+            throw new ActionParamsException("Invalid parameter value for key: " + StatisticsHelper.PARAM_SELECTORS + " - expected JSON object");
```

Comment:
`PARAM_DATA` is used to retrieve the data, but `PARAM_SELECTORS` is used in the error message.